### PR TITLE
Make CSR definitions scattered

### LINKED ISSUE #410: CSRs should use scattered functions
The CSR code is a bit odd. For example `readCSR()` uses a `match`:

```
function readCSR csr : csreg -> xlenbits = {
  let res : xlenbits =
  match (csr, sizeof(xlen)) {
    /* machine mode */
    (0xF11,  _) => zero_extend(mvendorid),
...
```

but *also* a scattered function for extensions.

```
...

    _           => /* check extensions */
                   match ext_read_CSR(csr) {
                     Some(res) => res,
                     None()    => { print_bits("unhandled read to CSR ", csr);
                                    zero_extend(0x0) }
                   }
```

This is silly. It should just work like the scattered functions for instructions - everything uses scattered functions, and there's a `riscv_csrs_begin.sail` and `riscv_csrs_end.sail`, and you add new CSRs just by inserting extra files.

### PR DESCRIPTION
**Pull request by Timmmm**
Resolve the strange split between `read_CSR` (which is not `scattered`) and `ext_read_CSR` (which is). Same for `write_CSR` and `is_CSR_defined`.

I changed the return type of `read_CSR` from `option(xlenbits)` to `xlenbits` since the code must already check `is_CSR_defined` before calling `read_CSR`. The only function that returned `None()` was the `seed` CSR in `write_seed_csr`, which actually meant you would get a weird "Unhandled write to CSR" if you wrote to `seed`.

I renamed & reordered the files slightly to make the scattered mapping work, but I haven't moved any of the actual definitions yet. In future we should actually scatter them.

Fixes #410

### CODE REVIEW

**Review by arichardson — COMMENTED** [review]
Not reviewed this in detail - but the result is much cleaner IMO. Having two methods for the same thing is weird, and I believe extensions can just override the existing definitions by placing their clause first.

File: model/riscv_insts_zicsr.sail
Comment (arichardson) [inline model/riscv_insts_zicsr.sail:189]: Since there is only one caller this seems fine but I wonder if it makes sense to have a logging variant.

In the future it might be nice to have logging for all implicit CSR reads and writes, but not for this PR.
I was thinking of maybe something like a scattered enum for all the CSRs so you don't have to hardcode the number and then using something like
`let mtvec = CSR(CSR_MTVEC)` could log the implicit read.

**Review by arichardson — APPROVED** [review]
Looked at the whole diff now. LGTM. Removing all the `Some(` is quite nice as well.

File: model/riscv_insts_zicsr.sail
Comment (Timmmm) [reply model/riscv_insts_zicsr.sail:189]: Yeah we have had to put stuff like `trackCSRWrite(csr_name_map("mstatus"))` in a few places in our code. A scattered enum might make more sense. It would also make the `write_CSR`/`read_CSR` a lot code easier to read.

Something like

```
scattered enum CSR

enum clause CSR = MSTATUS
enum clause CSR = MTVEC
...

scattered mapping csr_name_map : CSR <-> string

mapping class csr_name_map = MSTATUS <-> "mstatus"
mapping class csr_name_map = MTVEC <-> "mtvec"
...

scattered mapping csr_index : CSR <-> bits(12)

mapping clause csr_index = MSTATUS <-> 0x123
mapping clause csr_index = MTVEC <-> 0x456
...

function execute(CSR... = {
  let index : bits(12) = ...
  if not(csr_index_matches(index)) ...
  let csr = csr_index(index);
  write_CSR(csr, ...)
```

I'm not sure I quite follow what you mean by "a logging variant"?

Comment (arichardson) [reply model/riscv_insts_zicsr.sail:189]: I meant I am not sure if we want every read/write of a CSR to be emitted in the logfile since that could end up being really verbose.

Comment (arichardson) [reply model/riscv_insts_zicsr.sail:189]: And by logging variant I meant a version of write_CSR/read_CSR that logs the access as opposed to doing it in the caller.

**Comment by billmcspadden-riscv** [comment]
Minstret and timers are overkill.

File: model/riscv_insts_zicsr.sail
Comment (Timmmm) [reply model/riscv_insts_zicsr.sail:189]: Ah I see, that makes sense.

Comment (Timmmm) [reply model/riscv_insts_zicsr.sail:189]: (But I think we should do it in a future PR.)

Comment (arichardson) [reply model/riscv_insts_zicsr.sail:189]: agreed, no need to do it here.

**Comment by Timmmm** [comment]
I'll merge this in a few days if nobody objects. I think it's uncontroversial.

**Review by Alasdair — APPROVED** [review]
Once this is rebased I think it can be merged

**Pull request merged**
